src/testing/scanpy/_pytest/fixtures/__init__.py

```diff
@@ -53,10 +54,17 @@ def showwarning(message, category, filename, lineno, file=None, line=None) -> No
         else:
             showwarning_orig(message, category, filename, lineno, file, line)
 
+    # ignore plt.show() warning only in doctests.
+    warnings.filterwarnings(
+        "ignore",
+        r"FigureCanvasAgg is non-interactive, and thus cannot be shown",
+        UserWarning,
+    )
     # make errors visible and the rest ignored
+    action_map = defaultdict(lambda: "ignore", error="default")
     warnings.filters = [
-        ("default", *rest) for action, *rest in warnings.filters if action == "error"
-    ] + [("ignore", None, Warning, None, 0)]
+        (action_map[action], *rest) for action, *rest in warnings.filters
+    ]
```

Comment:
this is much better now, as it will preserve the warning filters more accurately: any action apart from “ignore” and “error” will display a warning (once or more often), so this should keep the precedence intact while downgrading each warning one level.

ignoring the few warnings we let through or ignore in the rest of the test suite and displaying the vast majority of warnings we upgrade into errors in the rest of the test suite as doctest output.